[Callbacks] DOC Add the user documentation for callbacks (#33812)
scikit-learn/scikit-learn: doc/callbacks.rst

.. currentmodule:: sklearn.callback

> .. _callbacks:

Reviewer: It's the same ref as in the dev guide I believe. It may case referencing issues. Let's make them distinct

Callbacks
=========

.. note::

  The callback API is experimental, and is not yet implemented for all estimators.
  Please refer to the :ref:`list of callback compatible estimators
  <callback_compatible_estimators>` for more information. It may change without the
  usual deprecation cycle.

This guide demonstrates how to use scikit-learn's callbacks on compatible estimators. If
you are developing a scikit-learn compatible estimator or meta-estimator and want to
make it compatible with callbacks, you can check our related developer guide. If you
want to develop a callback to use with callback compatible estimators, you can refer
to the related section of the developer guide.

.. TODO add refs to callback developer doc

Generally speaking, a callback is a mechanism used to execute custom code at specific
points during an algorithm's execution. Callbacks provide a clean way to insert custom
logic like monitoring progress or metrics, without modifying the underlying algorithm.

In scikit-learn, callbacks are objects that can be registered to an estimator. During
the fitting of the estimator, the callbacks registered to it will have their methods

> called at specific steps of the fitting process.

Reviewer: I'd rather merge both paragraphs and be less generic, but directly define callbacks in the context of scikit-learn
```suggestion
In scikit-learn, callbacks are objects from the :mod:`callbacks` module that can be
registered on an estimator to insert custom logic like monitoring progress or metrics,
without modifying the underlying learning algorithm. The registered callbacks are
called at specific steps of the fitting process.
```

Callback registration examples

> ******************************

Reviewer: ```suggestion
Registering callbacks
*********************
```

The following snippet demonstrates how a :class:`~ProgressBar` callback can be
registered on a :class:`~sklearn.linear_model.LogisticRegression` estimator using its

> :meth:`~CallbackSupportMixin.set_callbacks` method::

Reviewer: ```suggestion
Estimators that support callbacks expose a :meth:`~CallbackSupportMixin.set_callbacks`
method to register callbacks on them. The following example shows how to register a :class:`~ProgressBar` callback on a :class:`~sklearn.linear_model.LogisticRegression`::
```

    >>> from sklearn.callback import ProgressBar
    >>> from sklearn.linear_model import LogisticRegression
    >>> estimator = LogisticRegression()
    >>> progress_bar = ProgressBar()

>     >>> estimator.set_callbacks(progress_bar)

Reviewer: ```suggestion
    >>> progress_bar = ProgressBar()
    >>> estimator = LogisticRegression()
    >>> estimator.set_callbacks(progress_bar)
```

>     >>> estimator.set_callbacks(progress_bar)

Reviewer: These code snippet are executed as doctests so this one currently breaks. I'd like to merge this PR before logistic regression though. So maybe add a doctest skip for now with a TODO to remove when logistic regression supports callbacks

Now that the progress bar is registered to the estimator, calling its `fit` method will

> display a bar monitoring the progress of the `fit` execution::

Reviewer: ```suggestion
Now that the progress bar is registered on the estimator, calling its `fit` method will
display a bar monitoring the progress of the `fit` execution::
```

Author: I have used "register to" in the dev doc, and I try to keep consistent, although some "register on" might remain here and there. I have a slight preference for "register to" rather than "register on" but I could live with the latter. Maybe this is the opportunity to explicitly decide on a formulation so we can have the same everywhere.

Reviewer: If I'm not mistaken "register to" is not correct in that context. It should be "register with" or "register on". We went for "register on" in most cases. 

FOr tyhe dev docs I believe I already fixed them all.

Author: It sounded better to my naive ear, but after looking it up it appears that "register to" must be followed by a verb, but "registered to" + noun is ok, though apparently in software doc "registered on" is the dominant phrasing.

    >>> from sklearn.datasets import load_iris
    >>> X, y = load_iris(return_X_y=True)
    >>> estimator.fit(X, y)
    LogisticRegression - fit ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:03:12

Multiple callbacks can be registered to the same estimator by providing a list to
:meth:`~CallbackSupportMixin.set_callbacks`, for example a :class:`~ScoringMonitor`

> callback can be registered in addition to the :class:`~ProgressBar` one::

Reviewer: after https://github.com/scikit-learn/scikit-learn/pull/33799 you don't pass a list. I'd just say "Multiple callbacks can be registered to the same estimator"

Author: Ah yes I had noticed the change and modified the snippet but I forgot to change the text.

    >>> from sklearn.callback import ScoringMonitor
    >>> scoring_monitor = ScoringMonitor(scoring="precision")
    >>> estimator.set_callbacks(progress_bar, scoring_monitor)

Usage with meta-estimators
**************************

:term:`Meta-estimators` are estimator that, in their `fit`, will fit other estimators

> (designated as sub-estimators). For example a

Reviewer: no need to redefine here what meta estimators are since you already link to the glossary

:class:`~sklearn.model_selection.GridSearchCV` can be used to find the best parameters
for a :class:`~sklearn.linear_model.LogisticRegression` by fitting multiple instances of
that sub-estimator. Callbacks will behave differently in such estimator composition,

> depending whether the callback is an :class:`~AutoPropagatedCallback` or not.

Reviewer: It's not just that they behave differently but that they have different pusposes and aren't expected to be set on the same estimators. We should add a bit more details here.

Auto-propagated callbacks

> -------------------------

Reviewer: I'd rather put the regular callbacks first. Also I don't like referring to them as non-auto-propagated. We can call them regular imo

An :class:`~AutoPropagatedCallback`, such as :class:`~ProgressBar`, is a callback that
can be propagated to sub-estimators when registered to a meta-estimator. Thus its
methods can be called in the `fit` executions of both the meta-estimator and the
sub-estimators, allowing to aggregate information about the `fit` process of both types

> of estimators. If the sub-estimators are also meta-estimator, the auto-propageted

Reviewer: ```suggestion
of estimators. If the sub-estimators are also meta-estimator, the auto-propagated
```

callback can be propagated to the sub-sub-estimators too, and so on. This callback

> propagation is controlled through a `max_propagation_depth` argument of the callback.

Reviewer: ```suggestion
propagation is controlled through the `max_propagation_depth` argument of the callback.
```

This argument indicates the estimator depth up to which the callback will be propagated
in an estimator composition.

Auto-propagated callbacks must be registered to the top level meta-estimator of an
estimator composition. Trying to register it to any sub-estimator will raise an error

> when fitting the top level meta-estimator.

Reviewer: That's not entirely true. If the meta-estimator doesn't support callbacks you can. Let's put all that in a dropdown

AutoPropagatedCallback examples

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: Being the only subsection I'd rather flatten it

In the case of a
:class:`~sklearn.model_selection.GridSearchCV` applied on a
:class:`~sklearn.linear_model.LogisticRegression`, the auto-propagated callback

> :class:`~ProgressBar` must be registered to the :class:`~sklearn.search.GridSearchCV`::

Reviewer: ```suggestion
:class:`~ProgressBar` must be registered to the :class:`~sklearn.model_selection.GridSearchCV`::
```

    >>> from sklearn.model_selection import GridSearchCV

>     >>> parameters = {"l1_ratio": [0, 1], "fit_intercept": [True, False]}

Reviewer: `C` is the most common param to tune for logistic regression :)

    >>> sub_estimator = LogisticRegression()
    >>> meta_estimator = GridSearchCV(sub_estimator, parameters)
    >>> meta_estimator.set_callbacks(ProgressBar(max_propagation_depth=1))
    >>> meta_estimator.fit(X, y)
    #TODO add printed output

Changing the `max_propagation_depth` argument to 0 will make the auto-propagated

> callback only be registered to the top-level estimtor, in that case it means that only

Reviewer: ```suggestion
callback only be registered to the top-level estimator, in that case it means that only
```

one progressbar for the :class:`~sklearn.model_selection.GridSearchCV` will be
displayed::

    >>> meta_estimator.set_callbacks(ProgressBar(max_propagation_depth=0))
    >>> meta_estimator.fit(X, y)
    #TODO add printed output

Trying to register the auto-propagated callback on the sub-estimator will raise an
error::

    >>> sub_estimator = LogisticRegression().set_callbacks(ProgressBar())
    >>> meta_estimator = GridSearchCV(sub_estimator, parameters)
    >>> meta_estimator.fit(X, y)

>     #TODO add printed output

Reviewer: I think that just mentionning the behavior in the dropdown is enough. Let's give examples on what works only

Non auto-propagated callbacks
-----------------------------

A non auto-propagated callback, such as :class:`~ScoringMonitor`, can be registered to
any meta or sub-estimator in an estimator composition, and will be only invoked in the

> `fit` of the estimators it was registered to. If a sub-estimator with a non

Reviewer: while true, it's not really its purpose. I'd rather phrase it differently. Like they are usually expected to be registered on an estimator to gather quantities of the learning part. We should phrase it better but you get the idea

auto-propagated callback is fitted multiple times by a meta-estimator, or if it is
cloned by a meta-estimator and the clones get fitted, the callback will be invoked in
each one of these `fit` executions.

Non auto-propagated examples

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: same let's not make it a sub section since it's the only one

In the case of a :class:`~sklearn.model_selection.GridSearchCV` applied on a
:class:`~sklearn.linear_model.LogisticRegression`, the non auto-propagated callback
:class:`~ScoringMonitor` can be registered to the
:class:`~sklearn.linear_model.LogisticRegression`::

    >>> scoring = ScoringMonitor(scoring="precision")
    >>> sub_estimator = LogisticRegression().set_callbacks(scoring)
    >>> meta_estimator = GridSearchCV(sub_estimator, parameters)
    >>> meta_estimator.fit()
    >>> scoring.get_logs()
    #TODO add printed output, or maybe use the logs in a plot

The non auto-propagated callback can also be registered to the

> :class:`~sklearn.search.GridSearchCV`::

Reviewer: ```suggestion
:class:`~sklearn.model_selection.GridSearchCV`::
```

    >>> sub_estimator = LogisticRegression()
    >>> meta_estimator = GridSearchCV(sub_estimator, parameters).set_callbacks(scoring)
    >>> meta_estimator.fit()
    >>> scoring.get_logs()

>     #TODO add printed output, or maybe use the logs in a plot

Reviewer: I would not put that example. The scoring monitor is useless on a grid search

Or it can be registered to both the :class:`~sklearn.model_selection.GridSearchCV` and
the :class:`~sklearn.linear_model.LogisticRegression`::

    >>> sub_estimator = LogisticRegression().set_callbacks(scoring)
    >>> meta_estimator = GridSearchCV(sub_estimator, parameters).set_callbacks(scoring)
    >>> meta_estimator.fit()
    >>> scoring.get_logs()

>     #TODO add printed output, or maybe use the logs in a plot

Reviewer: same

Scikit-learn's built-in callbacks

> *********************************

Reviewer: before this section, I think that we need to add a section to define tasks and explain that callbacks are called at the beginning and end of each task. This is because the resul of some callbacks will mention tasks (e.g. Scoring monitor's log has task_id, task_path, parent_task_path, ...)

We should also briefly mention that tasks have a tree structure that is usually reflected in the monitored quantities of a callback, and link to the task tree dropdown of the dev guide.

The built-in callbacks currently available in scikit-learn are the following:

============================== =========================================================
Callback                       Description

============================== =========================================================
:class:`~ProgressBar`          Display progress bars.

> :class:`~ScoringMonitor`       Log the a scoring metric at the end of each task during

Reviewer: ```suggestion
:class:`~ScoringMonitor`       Log a scoring metric at the end of each task during
```
